# CAP5610 HW4 Task 2
## Robert Forristall

In [1]:
import numpy as np
import pandas as pd

train_x = pd.read_csv('data.csv')
train_y = pd.read_csv('label.csv')

# Data Visualization

In [2]:
train_x.describe()

0     0.1     0.2     0.3     0.4     0.5     0.6     0.7     0.8  \
count  9999.0  9999.0  9999.0  9999.0  9999.0  9999.0  9999.0  9999.0  9999.0   
mean      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
std       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
min       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
25%       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
50%       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
75%       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
max       0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

          0.9  ...        0.658        0.659        0.660        0.661  \
count  9999.0  ...  9999.000000  9999.000000  9999.000000  9999.000000   
mean      0.0  ...     0.179318     0.163616     0.052605     0.000600   
std       0.0  ...     5.674433     5.736359     2.420125     0.060003   
min       0.0  ...     0.000000     0.000000     0.000000     0.000000   
25%       0.0  ...     0.000000     0.000000     0.000000     0.000000   
50%       0.0  ...     0.000000     0.000000     0.000000     0.000000   
75%       0.0  ...     0.000000     0.000000     0.000000     0.000000   
max       0.0  ...   253.000000   253.000000   156.000000     6.000000   

        0.662   0.663   0.664   0.665   0.666   0.667  
count  9999.0  9999.0  9999.0  9999.0  9999.0  9999.0  
mean      0.0     0.0     0.0     0.0     0.0     0.0  
std       0.0     0.0     0.0     0.0     0.0     0.0  
min       0.0     0.0     0.0     0.0     0.0     0.0  
25%       0.0     0.0     0.0     0.0     0.0     0.0  
50%       0.0     0.0     0.0     0.0     0.0     0.0  
75%       0.0     0.0     0.0     0.0     0.0     0.0  
max       0.0     0.0     0.0     0.0     0.0     0.0  

[8 rows x 784 columns]

In [3]:
print(train_x.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Columns: 784 entries, 0 to 0.667
dtypes: int64(784)
memory usage: 59.8 MB
None


In [4]:
print(train_x.head())

   0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  ...  0.658  0.659  0.660  \
0  0    0    0    0    0    0    0    0    0    0  ...      0      0      0   
1  0    0    0    0    0    0    0    0    0    0  ...      0      0      0   
2  0    0    0    0    0    0    0    0    0    0  ...      0      0      0   
3  0    0    0    0    0    0    0    0    0    0  ...      0      0      0   
4  0    0    0    0    0    0    0    0    0    0  ...      0      0      0   

   0.661  0.662  0.663  0.664  0.665  0.666  0.667  
0      0      0      0      0      0      0      0  
1      0      0      0      0      0      0      0  
2      0      0      0      0      0      0      0  
3      0      0      0      0      0      0      0  
4      0      0      0      0      0      0      0  

[5 rows x 784 columns]


In [5]:
train_y.describe()

7
count  9999.000000
mean      4.443144
std       2.895897
min       0.000000
25%       2.000000
50%       4.000000
75%       7.000000
max       9.000000

In [6]:
print(train_y.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   7       9999 non-null   int64
dtypes: int64(1)
memory usage: 78.2 KB
None


In [7]:
print(train_y.head())

   7
0  2
1  1
2  0
3  4
4  1


# Q1
## Setup Distance Functions

In [8]:
def EuclideanDistance (previous_centroids, data):
    distance = np.zeros((len(previous_centroids),len(data)))
    for i in range(len(previous_centroids)):
        temp = []
        for point in data:
            temp.append(np.sqrt(np.sum(np.square(point-previous_centroids[i]))))
        distance[i] = temp
    return distance
def CosineDistance (previous_centroids, data):
    distance = np.zeros((len(previous_centroids),len(data)))
    for i in range(len(previous_centroids)):
        temp = []
        for point in data:
            # temp.append(1-(np.dot(previous_centroids[i], point)/(np.sqrt(np.sum(np.square(previous_centroids[i])))*np.sqrt(np.sum(np.square(point))))))
            temp.append(1 - (np.dot(previous_centroids[i],point)/(np.linalg.norm(previous_centroids[i])*np.linalg.norm(point))))
        distance[i] = temp
    return distance
def JarcardDistance (previous_centroids, data):
    distance = np.zeros((len(previous_centroids),len(data)))
    for i in range(len(previous_centroids)):
        temp = []
        for point in data:
            # min_arr = [min(x,y) for x,y in zip(previous_centroids[i], point)]
            # max_arr = [max(x,y) for x,y in zip(previous_centroids[i], point)]
            min_arr = np.minimum(previous_centroids[i], point)
            max_arr = np.maximum(previous_centroids[i], point)
            temp.append(1-(np.sum(min_arr)/np.sum(max_arr)))
        distance[i] = temp
    return distance

## Setup KMeans Function

In [9]:
np.random.seed(123)

def ShouldStop(iteration, centroids, old_centroids):
    if np.array_equal(old_centroids, centroids) or iteration == 100:
        return False
    return True

def getLabels (distance, centroids, data):
    labels = np.array([])
    for i in range(len(data)):
        index = np.where(distance == np.min(distance[:,i]))
        labels = np.append(labels, index[0][0])
    return labels

def CalculateCentroids (data, labels, previous_centroids, k, num_of_features):
    new_centroids = np.zeros((k, num_of_features))
    counts = [0]*k
    for i in range(len(data)):
        counts[labels[i]] += 1
        new_centroids[labels[i]] += data[i]
    for i in range(k):
        if counts[i] != 0:
            new_centroids[i] = new_centroids[i]/counts[i]
        else:
            new_centroids[i] = previous_centroids[i]
    return new_centroids

def KMeans (data, k_centroids, distance_method, stop_version, flag):
    
    num_of_features = len(data[0])
    current_centroids = np.zeros((k_centroids, num_of_features))
    for i in range(k_centroids):
        r = np.random.randint(0, num_of_features+1)
        current_centroids[i, :] = data[r]
    
    current_iteration = 0
    previous_centroids = np.zeros((k_centroids, num_of_features))
    labels = None
    old_sse = np.Inf
    sse = np.Inf
    
    if (stop_version == 1):
        while ShouldStop(current_iteration, current_centroids, previous_centroids):
            previous_centroids = current_centroids
            current_iteration += 1
            if (distance_method == 'euclidean'):
                distance = EuclideanDistance(previous_centroids, data)
            elif (distance_method == 'cosine'):
                distance = CosineDistance(previous_centroids, data)
            elif (distance_method == 'jarcard'):
                distance = JarcardDistance(previous_centroids, data)
            else:
                print('Invalid distacne metric: Please use eulidean, cosine, or jarcard')
                return
            labels = getLabels(distance, previous_centroids, data)
            current_centroids = CalculateCentroids(data, labels.astype(int), previous_centroids, k_centroids, num_of_features)
    elif (stop_version == 2):
        while ShouldStopV2(current_iteration, current_centroids, previous_centroids, old_sse, sse):
            old_sse = sse
            previous_centroids = current_centroids
            current_iteration += 1
            if (distance_method == 'euclidean'):
                distance = EuclideanDistance(previous_centroids, data)
            elif (distance_method == 'cosine'):
                distance = CosineDistance(previous_centroids, data)
            elif (distance_method == 'jarcard'):
                distance = JarcardDistance(previous_centroids, data)
            else:
                print('Invalid distacne metric: Please use eulidean, cosine, or jarcard')
                return
            labels = getLabels(distance, previous_centroids, data)
            current_centroids = CalculateCentroids(data, labels.astype(int), previous_centroids, k_centroids, num_of_features)
            sse = SSE(current_centroids, labels.astype(int), data)
    else:
        while ShouldStopV3(current_iteration, current_centroids, previous_centroids, old_sse, sse, flag):
            old_sse = sse
            previous_centroids = current_centroids
            current_iteration += 1
            if (distance_method == 'euclidean'):
                distance = EuclideanDistance(previous_centroids, data)
            elif (distance_method == 'cosine'):
                distance = CosineDistance(previous_centroids, data)
            elif (distance_method == 'jarcard'):
                distance = JarcardDistance(previous_centroids, data)
            else:
                print('Invalid distacne metric: Please use eulidean, cosine, or jarcard')
                return
            labels = getLabels(distance, previous_centroids, data)
            current_centroids = CalculateCentroids(data, labels.astype(int), previous_centroids, k_centroids, num_of_features)
            sse = SSE(current_centroids, labels.astype(int), data)
    print(distance_method, ' ', 'total interations: ', current_iteration)
    return current_centroids, labels
        
            


## Run KMeans Functions

In [10]:
# Euclidean Distance
euclidean_centroids, euclidean_labels = KMeans(train_x.to_numpy(), 10, 'euclidean', 1, 1)

# Cosine Distance
cosine_centroids, cosine_labels = KMeans(train_x.to_numpy(), 10, 'cosine', 1, 1)

# Jarcard Distance
jarcard_centroids, jarcard_labels = KMeans(train_x.to_numpy(), 10, 'jarcard', 1, 1)

euclidean   total interations:  71
cosine   total interations:  85
jarcard   total interations:  45


## Check SSE for each distance model

In [11]:
def SSE (centroids, labels, data):
    rv = 0
    for i in range(len(data)):
        index = labels[i]
        rv += np.sum(np.square(data[i]-centroids[index]))
    return rv
euclidean_sse = SSE (euclidean_centroids, euclidean_labels.astype(int), train_x.to_numpy())
cosine_sse = SSE (cosine_centroids, cosine_labels.astype(int), train_x.to_numpy())
jarcard_sse = SSE (jarcard_centroids, jarcard_labels.astype(int), train_x.to_numpy())

print('Euclidean SSE: ', euclidean_sse)
print('Cosine SSE: ', cosine_sse)
print('Jarcard SSE: ', jarcard_sse)
        

Euclidean SSE:  25318301268.749744
Cosine SSE:  25411581642.345028
Jarcard SSE:  25483177207.831245


## Evaluating the SSE values

Based on the following SSE values obtained during testing:
- Euclidean SSE:  25318301268.749744
- Cosine SSE:  25411581642.345028
- Jarcard SSE:  25483177207.831245

It would seem that the Euclidean method has the least error since it had the lowest SSE score.

# Q2
## Comparing the accuracies

In [12]:
def AccuracyCheck (centroids, centroid_labels, true_labels):
    hashmap = {}
    rv = 0
    for i in range(len(centroid_labels)):
        if centroid_labels[i] in hashmap:
            hashmap[centroid_labels[i]] = np.append(hashmap[centroid_labels[i]], true_labels['7'][i])
        else: 
            hashmap[centroid_labels[i]] = np.array([true_labels['7'][i]])
    for i in range(len(centroids)):
        if i in hashmap:
            vote = np.bincount(hashmap[i]).argmax()
            rv += np.count_nonzero(hashmap[i] == vote)
    return rv/len(true_labels)

euclidean_acc = AccuracyCheck(euclidean_centroids, euclidean_labels, train_y)
cosine_acc = AccuracyCheck(cosine_centroids, cosine_labels, train_y)
jarcard_acc = AccuracyCheck(jarcard_centroids, jarcard_labels, train_y)

print ('Euclidean Accuracy: ', euclidean_acc)
print('Cosine Accuracy: ', cosine_acc)
print('Jarcard Accuracy: ', jarcard_acc)

Euclidean Accuracy:  0.5942594259425943
Cosine Accuracy:  0.6140614061406141
Jarcard Accuracy:  0.6233623362336234


## Analyzing accuracies

Based on the following accuracy scores obtained through testing:
- Euclidean Accuracy:  0.5942594259425943
- Cosine Accuracy:  0.6140614061406141
- Jarcard Accuracy:  0.6233623362336234

It would seem that the Jarcard is the most accurate model based on the results of testing

# Q3
## Modifiying stop criteria to include SSE

In [13]:
def ShouldStopV2 (iteration, centroids, old_centroids, old_sse, sse):
    if (iteration != 0):
        if np.array_equal(old_centroids, centroids) or iteration == 100 or sse >= old_sse:
            return False
    return True

# Euclidean Distance
euclidean_centroids, euclidean_labels = KMeans(train_x.to_numpy(), 10, 'euclidean', 2, 1)

# Cosine Distance
cosine_centroids, cosine_labels = KMeans(train_x.to_numpy(), 10, 'cosine', 2, 1)

# Jarcard Distance
jarcard_centroids, jarcard_labels = KMeans(train_x.to_numpy(), 10, 'jarcard', 2, 1)

print('done')


euclidean   total interations:  93
cosine   total interations:  27
jarcard   total interations:  38
done


## Comparing iterations/times of models with modified stop criteria
Based on the results obtained during testing:
- Iterations till convergence:
    - Euclidean: 93
    - Cosine: 27
    - Jarcard: 38
- Times: Each function runs around the same amount of time per iteration so the shortest time was Cosine followed by Jarcard and lastly the slowest was euclidean due to the massive number of iterations this attempt took to converge

# Q4
## Modify stop conditions to consider only one terminating condition

In [14]:
def ShouldStopV3 (iteration, centroids, old_centroids, old_sse, sse, flag):
    if (iteration != 0):
        if (flag == 1):
            if np.array_equal(old_centroids, centroids):
                return False
        elif (flag == 2):
            if (sse >= old_sse):
                return False
        elif (flag == 3):
            if (iteration == 100):
                return False
    return True

# Euclidean Distance, includes what condition was used to stop
euclidean_centroids_position, euclidean_labels_position = KMeans(train_x.to_numpy(), 10, 'euclidean', 3, 1)
euclidean_centroids_sse, euclidean_labels_sse = KMeans(train_x.to_numpy(), 10, 'euclidean', 3, 2)
euclidean_centroids_iteration, euclidean_labels_iteration = KMeans(train_x.to_numpy(), 10, 'euclidean', 3, 3)

# Cosine Distance, includes what condition was used to stop
cosine_centroids_position, cosine_labels_position = KMeans(train_x.to_numpy(), 10, 'cosine', 3, 1)
cosine_centroids_sse, cosine_labels_sse = KMeans(train_x.to_numpy(), 10, 'cosine', 3, 2)
cosine_centroids_iteration, cosine_labels_iteration = KMeans(train_x.to_numpy(), 10, 'cosine', 3, 3)

# Jarcard Distance, includes what condition was used to stop
jarcard_centroids_position, jarcard_labels_position = KMeans(train_x.to_numpy(), 10, 'jarcard', 3, 1)
jarcard_centroids_sse, jarcard_labels_sse = KMeans(train_x.to_numpy(), 10, 'jarcard', 3, 2)
jarcard_centroids_iteration, jarcard_labels_iteration = KMeans(train_x.to_numpy(), 10, 'jarcard', 3, 3)


euclidean   total interations:  38
euclidean   total interations:  36
euclidean   total interations:  100
cosine   total interations:  59
cosine   total interations:  53
cosine   total interations:  100
jarcard   total interations:  60
jarcard   total interations:  47
jarcard   total interations:  100


## Calculate SSE for each model

In [15]:
# Euclidean distance model SSEs
euclidean_sse_position = SSE (euclidean_centroids_position, euclidean_labels_position.astype(int), train_x.to_numpy())
euclidean_sse_sse = SSE (euclidean_centroids_sse, euclidean_labels_sse.astype(int), train_x.to_numpy())
euclidean_sse_iteration = SSE (euclidean_centroids_iteration, euclidean_labels_iteration.astype(int), train_x.to_numpy())

# Cosine distance model SSEs
cosine_sse_position = SSE (cosine_centroids_position, cosine_labels_position.astype(int), train_x.to_numpy())
cosine_sse_sse = SSE (cosine_centroids_sse, cosine_labels_sse.astype(int), train_x.to_numpy())
cosine_sse_iteration = SSE (cosine_centroids_iteration, cosine_labels_iteration.astype(int), train_x.to_numpy())

# Jarcard distance model SSEs
jarcard_sse_position = SSE (jarcard_centroids_position, jarcard_labels_position.astype(int), train_x.to_numpy())
jarcard_sse_sse = SSE (jarcard_centroids_sse, jarcard_labels_sse.astype(int), train_x.to_numpy())
jarcard_sse_iteration = SSE (jarcard_centroids_iteration, jarcard_labels_iteration.astype(int), train_x.to_numpy())

print('Euclidean SSEs wrt stop criteria: ')
print ('Centroids converge: ', euclidean_sse_position)
print('New SSE is larger than previous SSE: ', euclidean_sse_sse)
print('Reached max number of iterations (100): ', euclidean_sse_iteration)

print('Cosine SSEs wrt stop criteria: ')
print ('Centroids converge: ', cosine_sse_position)
print('New SSE is larger than previous SSE: ', cosine_sse_sse)
print('Reached max number of iterations (100): ', cosine_sse_iteration)

print('Jarcard SSEs wrt stop criteria: ')
print ('Centroids converge: ', jarcard_sse_position)
print('New SSE is larger than previous SSE: ', jarcard_sse_sse)
print('Reached max number of iterations (100): ', jarcard_sse_iteration)

Euclidean SSEs wrt stop criteria: 
Centroids converge:  25321994318.21326
New SSE is larger than previous SSE:  25495249192.75157
Reached max number of iterations (100):  25404775206.395275
Cosine SSEs wrt stop criteria: 
Centroids converge:  25468173815.67823
New SSE is larger than previous SSE:  25425454489.886997
Reached max number of iterations (100):  25416733862.67944
Jarcard SSEs wrt stop criteria: 
Centroids converge:  25415230917.736603
New SSE is larger than previous SSE:  25434179158.75361
Reached max number of iterations (100):  25587905239.143787


# Compare SSE wrt stop conditions

Based on the results obtained during testing:
- Euclidean SSEs wrt stop criteria: 
    - Centroids converge:  25429959337.949955
    - New SSE is larger than previous SSE:  25407016682.77918
    - Reached max number of iterations (100):  25321974749.58356
- Cosine SSEs wrt stop criteria: 
    - Centroids converge:  25555029379.927616
    - New SSE is larger than previous SSE:  25610084771.580967
    - Reached max number of iterations (100):  25561511207.494164
- Jarcard SSEs wrt stop criteria: 
    - Centroids converge:  25414464925.44861
    - New SSE is larger than previous SSE:  25446594079.129356
    - Reached max number of iterations (100):  25451185367.17009
    
The best models determined by SSE wrt to the different stop conditions:
- Centroids converge: Jarcard
- New SSE is larger than previous SSE: Euclidean
- Reached max number of iterations (100): Euclidean

Thus the best overall KMeans model as found from this analysis is Euclidean

# Q5
## Concluding observations and algorithm analysis

In conclusion to this algorithmic analysis it would seem that overall the most consistent and reliable distance metric for the KMeans model is difficult to determine. Each of these metrics had tests that they appeared to be the best at; euclidean distance did the best in regard to the SSE metric both overall and with respect to the majority of the stop conditions, meanwhile jarcard had the best accuracy score when compared to the label set while euclidean had the lowest accuracy score. The cosine distance metric seems to be a sort of compensation between the two with an overall SSE better than jarcard but worse than euclidean and accuracy being better than euclidean but worse than jarcard.

In the future other distance metrics such as manhattan distance will have to be tested on the data set to see if any better performance can be obtained. Other methods than the one used here for determining the initial clusters will likely also help improve the performance of this model since initial centroids could be selected in strategic positions rather than just using a random point in the data set.